### PR TITLE
Use existing coverage reports

### PR DESCRIPTION
**Pull request by borNfreee**
Fixes https://github.com/infection/infection/issues/94
- [x] Allow to provide the path to folder with existing coverage report(s)
- [x] Disable Xdebug for Initial tests run if existing coverage provided
- [x] Documentation (https://github.com/infection/site/pull/20)
- [x] Reuse coverage on Travis to speed up builds

### CODE REVIEW

**Comment by BackEndTea** [comment]
Could you add this option to travis? should save us some time during the build

File: src/TestFramework/Factory.php [hunk 1]
```diff
@@ -27,7 +27,7 @@
     /**
      * @var string
      */
-    private $tempDir;
+    private $tmpDir;
```
Comment (BackEndTea) [inline src/TestFramework/Factory.php]: I like that we are making this consistent all over the code base, but the diff of this PR is already quite big, so maybe this should be moved to a separate PR. That would make it a bit easier to see exactly what changed.

Comment (borNfreee) [reply src/TestFramework/Factory.php]: I will spend more time creating separate PR for these 2 files than review requires :) leaving it here

**Comment by borNfreee** [comment]
> Could you add this option to travis? should save us some time during the build

done

File: app/container.php [hunk 2]
```diff
@@ -56,12 +56,20 @@
...
 
-$c['coverage.dir.phpunit'] = function (Container $c): string {
-    return $c['tmp.dir'] . '/' . CodeCoverageData::PHP_UNIT_COVERAGE_DIR;
+$c['tmp.dir'] = function (Container $c): string {
+    return $c['tmp.dir.creator']->createAndGet($c['infection.config']->getTmpDir());
 };
 
-$c['coverage.dir.phpspec'] = function (Container $c): string {
-    return $c['tmp.dir'] . '/' . CodeCoverageData::PHP_SPEC_COVERAGE_DIR;
+$c['coverage.dir.phpunit'] = function (Container $c) {
```
Comment (sidz) [inline app/container.php]: is this service using elsewhere? can't find any usage

p.s. the same for phpspec

File: src/Command/InfectionCommand.php [hunk 3]
```diff
@@ -103,6 +104,13 @@ protected function configure()
                 InputOption::VALUE_REQUIRED,
                 'Custom configuration file path'
             )
+            ->addOption(
+                'coverage',
+                null,
+                InputOption::VALUE_REQUIRED,
```
Comment (sidz) [inline src/Command/InfectionCommand.php]: I'd say that this option shouldn't be mandatory. Can we do it optional?

File: app/container.php (hunk 2 above)
Comment (borNfreee) [reply app/container.php]: https://github.com/infection/infection/blob/master/src/Command/InfectionCommand.php#L278

File: src/Console/Application.php [hunk 4]
```diff
@@ -106,18 +107,38 @@ private function buildDynamicDependencies(InputInterface $input)
...
+    private function getExistingCoveragePath(InputInterface $input, string $configLocation): string
+    {
+        $existingCoveragePath = $input->getOption('coverage');
+
+        if ($existingCoveragePath === '') {
+            return '';
+        }
+
+        /** @var Filesystem $fileSystem */
+        $fileSystem = $this->container['filesystem'];
```
Comment (sidz) [inline src/Console/Application.php]: can we add filesystem as argument?

File: src/TestFramework/PhpUnit/Coverage/CoverageXmlParser.php [hunk 5]
```diff
@@ -134,7 +134,8 @@ private function getSourceFilePath(\DOMXPath $xPath, string $relativeCoverageFil
             return $realPath;
         }
 
-        throw new \Exception('Source file was not found');
+        throw new \Exception(
```
Comment (sidz) [inline src/TestFramework/PhpUnit/Coverage/CoverageXmlParser.php]: can we create smth like CoverageException? or put this message into CoverageDoesNotExistException?

File: src/Command/InfectionCommand.php (hunk 3 above)
Comment (borNfreee) [reply src/Command/InfectionCommand.php]: This parameter means that the value of the option is required, not for the option itself.

So `--coverage=dir` is allowed, but `--coverage` is not

File: src/Console/Application.php [hunk 6]
```diff
@@ -106,18 +107,38 @@ private function buildDynamicDependencies(InputInterface $input)
...
+
+            return $existingCoveragePath ?: $c['tmp.dir'];
         };
     }
+
+    private function getExistingCoveragePath(InputInterface $input, string $configLocation): string
+    {
+        $existingCoveragePath = $input->getOption('coverage');
+
+        if ($existingCoveragePath === '') {
```
Comment (sidz) [inline src/Console/Application.php]: do we need to trim this value?

File: src/TestFramework/PhpSpec/Config/InitialYamlConfiguration.php [hunk 7]
```diff
@@ -8,23 +8,68 @@
...
 use Symfony\Component\Yaml\Yaml;
 
 class InitialYamlConfiguration extends AbstractYamlConfiguration
 {
+    /**
+     * @var bool
+     */
+    private $skipCoverage;
+
+    public function __construct(string $tempDirectory, array $parsedYaml, bool $skipCoverage)
```
Comment (sidz) [inline src/TestFramework/PhpSpec/Config/InitialYamlConfiguration.php]: tmpDir?

File: src/Command/InfectionCommand.php [hunk 8]
```diff
@@ -103,6 +104,13 @@ protected function configure()
                 InputOption::VALUE_REQUIRED,
                 'Custom configuration file path'
             )
+            ->addOption(
+                'coverage',
```
Comment (sidz) [inline src/Command/InfectionCommand.php]: if it's path maybe we should call it `coverage-path` ?

Comment (borNfreee) [reply src/Command/InfectionCommand.php]: I thought the same but made it consistent with the `--configuration` options https://infection.github.io/guide/command-line-options.html#configuration-or-c which also does not have `path` postfix

**Comment by BackEndTea** [comment]
This may be outside of the scope of this PR, but if someone uses the --coverage option, there shouldn't be any need for them to have xdebug enabled, or run the command through phpdbg

For exmple, lets say i use phpdbg to get the coverage of my project, and i don't have xdebug installed.

I would then like to be able to run infection without phpdbg, since that slows it down. And as far as i know, once we have the original coverage, there is no need for xdebug or phpdbg.

File: src/Console/Application.php [hunk 9]
```diff
@@ -106,18 +107,39 @@ private function buildDynamicDependencies(InputInterface $input)
...
-            return new InfectionConfig(json_decode($json), $c['filesystem'], $dir);
+            return new InfectionConfig(json_decode($json), $c['filesystem'], $configLocation);
         };
 
-        $this->container['tmp.dir'] = function (Container $c): string {
-            return $c['tmp.dir.creator']->createAndGet($c['infection.config']->getTmpDir());
+        $this->container['coverage.path'] = function (Container $c) use ($input): string {
+            $existingCoveragePath = $this->getExistingCoveragePath(
+                $input,
+                $c['infection.config']->getPhpUnitConfigDir(),
```
Comment (BackEndTea) [inline src/Console/Application.php]: Since the --coverage option is a command line option, its location should be relative to where the command is ran from, not the phpunit config location

File: src/Process/Builder/ProcessBuilder.php [hunk 10]
```diff
@@ -36,12 +36,13 @@ public function __construct(AbstractTestFrameworkAdapter $testFrameworkAdapter,
      * Creates process with enabled debugger as test framework is going to use in the code coverage.
      *
      * @param string $testFrameworkExtraOptions
+     * @param bool $skipCoverage
```
[the remark this answers is not in the document]
Comment (borNfreee) [reply src/Process/Builder/ProcessBuilder.php]: we don't use useless PHPDoc `@param` since PHP7 already provides information *unless* there is a string, describing the method purpose as in this case `Creates process with enabled debugger as test framework is going to use in the code coverage.`

Without `@param` here, my IDE complains and underlines the method with warnings. I think this is a pretty common practice, and in this repository as well.

**Comment by borNfreee** [comment]
> there shouldn't be any need for them to have xdebug enabled, or run the command through phpdbg

As I understand, it will be easier to do it when https://github.com/infection/infection/pull/173 gets merged since we will have information about passed options in `$input` object. 

@BackEndTea is there anything else to be addressed here?

**Comment by sidz** [comment]
@borNfreee unfortunately we don't have necessary information about passed options in the Application.
At least we can't do `$input->getOption` because parameters binding happens a bit later.

**Comment by borNfreee** [comment]
Thank you @sidz & @BackEndTea
